vlog4j-core/src/main/java/org/semanticweb/vlog4j/core/reasoner/KnowledgeBase.java

```diff
@@ -1,434 +1,559 @@
+	}
+
+	<T> List<T> getStatementsByType(final Class<T> type) {
+		final ExtractStatementsVisitor<T> visitor = new ExtractStatementsVisitor<>(type);
+		for (final Statement statement : this.statements) {
+			statement.accept(visitor);
+		}
+		return Collections.unmodifiableList(visitor.getExtractedStatements());
+	}
+
+	/**
+	 * Add a single fact to the internal data structures. It is assumed that it has
+	 * already been checked that this fact is not present yet.
+	 *
+	 * @param fact the fact to add
+	 */
+	void addFact(final Fact fact) {
+		final Predicate predicate = fact.getPredicate();
+		this.factsByPredicate.putIfAbsent(predicate, new HashSet<>());
+		this.factsByPredicate.get(predicate).add(fact);
+	}
+
+	/**
+	 * Removes a single fact from the internal data structure. It is assumed that it
+	 * has already been checked that this fact is already present.
+	 *
+	 * @param fact the fact to remove
+	 */
+	void removeFact(final Fact fact) {
+		final Predicate predicate = fact.getPredicate();
+		final Set<PositiveLiteral> facts = this.factsByPredicate.get(predicate);
+		facts.remove(fact);
+		if (facts.isEmpty()) {
+			this.factsByPredicate.remove(predicate);
+		}
+	}
+
+	/**
+	 * Returns all {@link Statement}s of this knowledge base.
+	 *
+	 * The result can be iterated over and will return statements in the original
+	 * order. The collection is read-only and cannot be modified to add or delete
+	 * statements.
+	 *
+	 * @return a collection of statements
+	 */
+	public Collection<Statement> getStatements() {
+		return Collections.unmodifiableCollection(this.statements);
+	}
+
+	@Override
+	public Iterator<Statement> iterator() {
+		return Collections.unmodifiableCollection(this.statements).iterator();
+	}
+
+	Map<Predicate, Set<PositiveLiteral>> getFactsByPredicate() {
+		return this.factsByPredicate;
+	}
+
+	/**
+	 * Interface for a method that parses the contents of a stream into a
+	 * KnowledgeBase.
+	 *
+	 * This is essentially
+	 * {@link org.semanticweb.vlog4j.parser.RuleParser#parseInto}, but we need to
+	 * avoid a circular dependency here -- this is also why we throw
+	 * {@link Exception} instead of
+	 * {@link org.semanticweb.vlog4j.parser.ParsingException}.
+	 */
+	@FunctionalInterface
+	public interface AdditionalInputParser {
+		KnowledgeBase parseInto(InputStream stream, KnowledgeBase kb) throws IOException, Exception;
+	}
+
+	/**
+	 * Import rules from a file.
+	 *
+	 * @param file          the file to import
+	 * @param parseFunction a function that transforms a {@link KnowledgeBase} using
+	 *                      the {@link InputStream}.
+	 *
+	 * @throws IOException              when reading {@code file} fails
+	 * @throws IllegalArgumentException when {@code file} is null or has already
+	 *                                  been imported
+	 * @throws RuntimeException         when parseFunction throws
+	 *
+	 * @return this
+	 */
+	public KnowledgeBase importRulesFile(File file, AdditionalInputParser parseFunction)
```

Comment:
why do we need to return a `KnowledgeBase `object, if we already have a reference to the same object, as we called importRulesFile on it?
I think it is clearer that this function modifies `this `if it returns void.